src/BinaryFile.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Php\Pie;
+
+use function hash_file;
+
+/**
+ * @internal This is not public API for PIE, so should not be depended upon unless you accept the risk of BC breaks
+ *
+ * @immutable
+ */
+final class BinaryFile
+{
+    private const HASH_TYPE_SHA256 = 'sha256';
```

Comment:
Why not sha3-256 for new stuff?

Reply:
See https://news-web.php.net/php.internals/124694 and specifically https://news.ycombinator.com/item?id=14469614.

Also SHA-256 comes with hardware support for PHP 8.4, so it's really blazing fast (php/php-src#15152).

Reply:
Ok, but if I read this correctly you have to do hash_hmac to prevent length extension attacks with SHA-256? How would that look here? Because in this file its just doing hash_file, is that safe then? I mean I don't think in this particular case the hashes are security-relevant, but just trying to understand.

Reply:
> Ok, but if I read this correctly you have to do hash_hmac to prevent length extension attacks with SHA-256?

You read that correctly, but this attack vector is not relevant to this use case. To my understanding we care only about integrity here. Length extension attacks break authentication: The attacker knows a valid hash and the corresponding payload, but does not know the signature key.

To give a simplified example, if the signature is created as follows:

```
$hash = hash('sha256', $secretKey . $payload);
```

Then an attacker would be able to determine the following hash, even without knowing the `$secretKey`:

```
$hash = hash('sha256', $secretKey . $payload . $extension);
```

This would allow the attacker to pretend that `$payload . $extension` is validly signed. This can be relevant for file formats where the last entry wins: For example a tar archive.

Reply:
> To my understanding we care only about integrity here

yep this is only for checking if the .so matches the one built by PIE (e.g. if it gets replaced externally), so doesn't really need to be secure as such, it is literally "has this thing changed" :)